### PR TITLE
Issue #2341 - added ZSink.drop

### LINKED ISSUE #2341: Add skip to ZSink/Sink
```scala
  // Skips n elements 
  def skip[A](n: Long): ZSink[Any, Unit, A, A, Unit] =
    ZSink.foldUntil[Long, A](0L, n)((s, _) => s + 1).flatMap {
      case m if n != m => ZSink.fail(())
      case _           => ZSink.succeed(())
    }

  // Gets the head of a stream
  def get[A]: ZSink[Any, Unit, A, A, A] =
    ZSink.foldUntil[Option[A], A](None, 1)((_, a) => Some(a)).flatMap {
      case None    => ZSink.fail(())
      case Some(a) => ZSink.succeed(a)
    }
```

I am not totally immersed into your ecosystem, so I did some testing in scalatest:
```scala
  property("test") {
    val rt = new DefaultRuntime {}

    {
      val r = Stream.fromIterable(List(1, 2, 3, 4, 5)).run(for {
        _ <- skip[Int](2)
        r <- get[Int]
      } yield r)
      assert(rt.unsafeRunSync(r).toEither.right.get == 3)
    }

    {
      val r = Stream.fromIterable(List(1, 2, 3)).run(for {
        _ <- skip[Int](2)
        r <- get[Int]
        _ <- skip[Int](1)
      } yield r)
      assert(rt.unsafeRunSync(r).toEither.isLeft)
    }

    {
      val r = Stream.fromIterable(List(1, 2, 3, 4)).run(for {
        _ <- skip[Int](2)
        r <- get[Int]
        _ <- skip[Int](1)
      } yield r)
      assert(rt.unsafeRunSync(r).toEither.right.get == 3)
    }
  }
```

### PR DESCRIPTION
**Pull request by pk044**
Closes #2341 .
It was supposed to be `skip` at first, but I changed it to `drop` to be more consistent with other abstractions (ZStream, ZStreamChunk etc.) - let me know if it's not a good idea and I'll change it :)

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 1]
```diff
@@ -1062,6 +1062,25 @@ object ZSink extends ZSinkPlatformSpecificConstructors with Serializable {
...
+        // Cast is redundant but required for Scala 2.11
+        val initial = sink.initial.map((_, 0L)).asInstanceOf[ZIO[R, E, this.State]]
+
+        def step(state: State, a: A) =
+          if (state._2 != n) UIO.succeed((state._1, state._2 + 1))
+          else sink.step(state._1, a).map((_, state._2))
+
+        def extract(state: State) = sink.extract(state._1)
+
+        def cont(state: State) = sink.cont(state._1)
```
Comment (vasilmkd) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala]: I would think the sink needs to continue for at least `n` iterations and then "switch over" to the underlying sink.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala [hunk 2]
```diff
@@ -1062,6 +1062,25 @@ object ZSink extends ZSinkPlatformSpecificConstructors with Serializable {
...
+     */
+    final def drop(n: Long): ZSink[R, E, A0, A, B] =
+      new ZSink[R, E, A0, A, B] {
+        type State = (sink.State, Long)
+
+        // Cast is redundant but required for Scala 2.11
+        val initial = sink.initial.map((_, 0L)).asInstanceOf[ZIO[R, E, this.State]]
+
+        def step(state: State, a: A) =
+          if (state._2 != n) UIO.succeed((state._1, state._2 + 1))
```
Comment (vasilmkd) [inline streams/shared/src/main/scala/zio/stream/ZSink.scala]: I'd prefer a more explicit `<` check here instead of `!=`.

File: streams/shared/src/main/scala/zio/stream/ZSink.scala (hunk 1 above)
Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/ZSink.scala]: or when both continue?
```scala
state._2 < n && sink.cont(state._1)
```

**Comment by vasilmkd** [comment]
Not really. This would stop exactly after `n` are dropped. After that point it should be only `sink.cont`.

**Comment by regiskuckaertz** [comment]
Hmm, but then it's as if `n` doesn't count at all, right? _If_ I understand what you mean:
```scala
if (state._2 < n)
  sink.cont(state._1)
else
  sink.cont(state._1)
```

**Comment by vasilmkd** [comment]
The way I understand this combinator is: it demands to drop `n` elements. To me, that means that until `n` elements have passed through the sink, it will not even consult the underlying one. So I would write the `cont` method as follows: `def cont(s: State): Boolean = s._2 < n || sink.cont(s._1)`

File: streams-tests/jvm/src/test/scala/zio/stream/SinkSpec.scala [hunk 3]
```diff
@@ -244,6 +244,24 @@ object SinkSpec extends ZIOBaseSpec {
...
+        testM("init error") {
+          val sink = initErrorSink.drop(1)
+          assertM(sinkIteration(sink, 1).either, isLeft(equalTo("Ouch")))
+        },
+        testM("step error") {
+          val sink = stepErrorSink.drop(1)
+          assertM(sinkIteration(sink, 1).either, isLeft(equalTo("Ouch")))
+        },
+        testM("extract error") {
+          val sink = extractErrorSink.drop(1)
```
Comment (regiskuckaertz) [inline streams-tests/jvm/src/test/scala/zio/stream/SinkSpec.scala]: Can you please add tests that drop more than one element? As also, one that drops zero element. 🙏
